### PR TITLE
Offer status

### PR DESCRIPTION
**Pull request by pldespaigne**
solve #6305

### CODE REVIEW

File: libs/utils/src/lib/pipes/total.pipe.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { getDeepValue } from './deep-key.pipe';
 @Pipe({ name: 'total' })
 export class TotalPipe implements PipeTransform {
   transform(data: unknown[], path: string): number {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/pipes/total.pipe.ts:6]: update type `transform(data?: unknown[] | null, path: string)` to reflect your code

File: apps/backend-functions/src/contracts.ts [hunk 2]
```diff
@@ -32,3 +34,48 @@ export async function onContractDelete(contractSnapshot: FirebaseFirestore.Docum
 
   console.log(`Contract ${contract.id} removed`);
 }
+
+
+export async function onContractUpdate(
+  change: Change<FirebaseFirestore.DocumentSnapshot>
+) {
+
+  const before = change.before;
```
Comment (GrandSchtroumpf) [inline apps/backend-functions/src/contracts.ts:43]: you can do : 
```typescript
const before = change.before?.data();
const after = change.after?.data();
if (!before || !after) {
  // Use HttpsError provides a better login system if I recall correctly
  throw new https.HttpsError('failed-precondition', 'Parameter "change" not found');
}
const isSale = before.type === after.type && before.type === 'sale';
const statusChanged = before.status !== after.status;
if (!isSale || !statusChanged) return;
...
```

File: apps/backend-functions/src/contracts.ts [hunk 3]
```diff
@@ -32,3 +34,48 @@ export async function onContractDelete(contractSnapshot: FirebaseFirestore.Docum
...
+  const isSale = contractBefore.type === contractAfter.type && contractBefore.type === 'sale' // contract is of type 'sale'
+  const statusHasChanged = contractBefore.status !== contractAfter.status // contract status has changed
+
+  if (isSale && statusHasChanged) {
+
+    const offerRef = db.doc(`offers/${contractAfter.offerId}`);
+    const offerSnap = await offerRef.get();
+    const offer = offerSnap.data() as Offer;
+
+    if (offer.status !== 'signed' && offer.status !== 'signing') {
```
Comment (GrandSchtroumpf) [inline apps/backend-functions/src/contracts.ts:63]: I like it better when you return after if : 
```typescript
if (offer.status === 'signed' || offer.status === 'signing') return;
```

File: apps/backend-functions/src/contracts.ts [hunk 4]
```diff
@@ -32,3 +34,48 @@ export async function onContractDelete(contractSnapshot: FirebaseFirestore.Docum
...
+
+      const offerContractsQuery = db.collection('contracts')
+        .where('offerId', '==', contractAfter.offerId)
+        .where('type', '==', 'sale');
+      const offerContractsSnap = await offerContractsQuery.get();
+
+      const contractsStatus: ContractStatus[] = offerContractsSnap.docs.map(doc => doc.data().status);
+
+      let newOfferStatus = offer.status;
+      newOfferStatus = contractsStatus.every(status => status !== 'accepted') ? 'pending' : newOfferStatus;
```
Comment (GrandSchtroumpf) [inline apps/backend-functions/src/contracts.ts:73]: You should remove this line. It does nothing. It's make the code too confusing.

File: apps/backend-functions/src/contracts.ts [hunk 5]
```diff
@@ -32,3 +34,48 @@ export async function onContractDelete(contractSnapshot: FirebaseFirestore.Docum
...
+      newOfferStatus = contractsStatus.some(status => status !== 'pending') ? 'negotiating' : newOfferStatus;
+      newOfferStatus = contractsStatus.every(status => status === 'accepted') ? 'accepted' : newOfferStatus;
+      newOfferStatus = contractsStatus.every(status => status === 'declined') ? 'declined' : newOfferStatus;
+
+      offerRef.update({ status: newOfferStatus });
```
Comment (GrandSchtroumpf) [inline apps/backend-functions/src/contracts.ts:78]: no need to update if status hasn't changed : 
```typescript
if (newOfferStatus === offer.status) return;
offerRef.update({ status: newOfferStatus });
```

**Review by GrandSchtroumpf — APPROVED** [review]

File: apps/backend-functions/src/contracts.ts (hunk 2 above)
Comment (pldespaigne) [reply apps/backend-functions/src/contracts.ts:43]: We cannot return something to the client as this function is invoked by an event in the db

**Pull request merged**